### ISSUE TITLE
`TypeVar` in union parameter with `type[T]` resolves to `Unknown` when argument subclasses `@final` generic class

### ISSUE THREAD

**Issue by cmp0xff** [astral-sh/ty#4364]
### Summary

When a generic function accepts a union containing `type[T]` (e.g. `Base[T] | type[T]`), `ty` fails to solve `T` if the argument passed is an instance of a subclass of a `@final` class (even when suppressed with `# ty: ignore[subclass-of-final-class]`), resolving `T` to `Unknown`.

Notably, calling `f(d: Base[T])` with a single parameter works as expected; the failure only occurs when
`Base[T]` is placed in a union with `type[T]`.

### Minimal Reproducible Example

https://play.ty.dev/12fb4bd8-2349-4f33-ae6c-c45a9295d1d7

```python
from typing import Generic, TypeVar, assert_type, final

T_co = TypeVar("T_co", covariant=True)
T = TypeVar("T")

@final
class Base(Generic[T_co]): ...

# Subclassing a final generic class (mirrors numpy.dtypes.*DType extending numpy.dtype)
class Leaf(Base[int]): ...  # ty: ignore[subclass-of-final-class]

def f_single(x: Base[T]) -> T:
    raise NotImplementedError

def f_union(x: Base[T] | type[T]) -> T:
    raise NotImplementedError

d = Leaf()

# 1. Single parameter: PASSES
assert_type(f_single(d), int)

# 2. Union parameter: FAILS
assert_type(f_union(d), int)
```

### ty Diagnostic

```
error[type-assertion-failure]: Argument does not have asserted type `int`
  --> test.py:22:1
    |
22 | assert_type(f_union(d), int)
    | ^^^^^^^^^^^^----------^^^^^^
    |             |
    |             Inferred type is `Unknown`
info: `int` and `Unknown` are not equivalent types
```

### Expected Behavior

T should be solved to int in f_union(d) matching the first branch Base[T], consistent with f_single(d) and other type checkers (mypy, pyright, pyrefly).

### Real-World Context

This pattern occurs in pandas-stubs / pandas when calling Series.to_numpy(dtype=np.dtypes.Int8DType()). Series.to_numpy defines dtype: np.dtype[GenericT] | SupportsDType[GenericT] | type[GenericT]. Because numpy.dtype is marked @final in NumPy stubs and Int8DType subclasses it, ty fails to solve GenericT and falls back to an untyped overload returning ndarray[..., dtype[Any]].

The numpy issue has been reported in numpy/numpy#32390.

### Version

playground

**Comment by carljm** [comment]
I think it should be valid for a type checker to assume that a class marked as `@final` does not have subclasses. This often allows type checkers to make more precise type inferences or improve performance; allowing type checkers to make such assumptions is part of the purpose of `@final`. So fundamentally I think the annotation approach here is invalid, and `Unknown` is a valid inference in that invalid scenario. (And I expect in general that libraries which subclass final classes are likely to encounter other typing issues as fallout from that.)

What ty is doing here is observing that `Leaf` cannot be a subclass of `Base` (because `Base` is final) and is not a subclass of `type`, therefore neither arm of the union can match, and there is no basis for inferring a type for `T`, so we fall back to `Unknown`.

That said, our behavior is inconsistent, because we do infer a precise type for `f_single(Leaf())`. The finality check here only happens in a union-specific filtering pass, which I expect to be removed as part of #3557. So in practice I expect we will start inferring a precise type for `f_union(Leaf())`.
